libs/datamodel/core/src/transform/ast_to_dml/db.rs

```diff
@@ -1,31 +1,151 @@
+#![allow(unused)]
+
 mod names;
+mod relations;
+mod types;
 
-use crate::{ast, diagnostics::Diagnostics};
+use self::relations::RelationField;
+use crate::{
+    ast::{self, FieldId, TopId},
+    diagnostics::{DatamodelError, Diagnostics},
+};
+use names::Names;
+use std::str::FromStr;
 
+/// Information gathered during schema validation. Each validation steps
+/// enriches the database with information that can be used in later steps.
 pub(crate) struct ParserDatabase<'a> {
     ast: &'a ast::SchemaAst,
-    names: names::Names<'a>,
+    names: Names<'a>,
+    types: types::Types,
+    relations: relations::Relations,
 }
 
 impl<'ast> ParserDatabase<'ast> {
-    pub(super) fn new(ast: &'ast ast::SchemaAst, diagnostics: &mut Diagnostics) -> Self {
-        let names = names::Names::new(ast, diagnostics);
+    pub(super) fn new(ast: &'ast ast::SchemaAst, diagnostics: &mut Diagnostics) -> Option<Self> {
+        let names = Names::new(ast, diagnostics);
+
+        if diagnostics.has_errors() {
+            return None;
+        }
+
+        let mut types = types::Types::default();
+        let mut relations = relations::Relations::default();
+
+        for (top_id, top) in ast.iter_tops() {
+            match top {
+                ast::Top::Type(type_alias) => {
+                    match field_type(type_alias, &names, ast) {
+                        Ok(FieldType::Scalar(scalar_field_type)) => {
+                            types.type_aliases.insert(top_id, scalar_field_type);
```

Comment:
```rust
types.insert_alias(top_id, scalar_field_type);
```
?

Reply:
I'm still figuring out the code organization here. Do we need these private `Types`, `Names`, `Relations` modules? If so, how encapsulated should they be? My current thinking is that they should be more like containers without logic, with the logic living in `ParserDatabase`.

Reply:
But we can definitely make this change.